File: 2025_go/09_movie.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type coord struct{ x, y int }
type seg struct{ fixed, lo, hi int }

func main() {
	fmt.Println("Sample A", solve09("inputs/09_sample.txt", false)) // 50
	fmt.Println("A", solve09("inputs/09_input.txt", false))         // 4776100539
	fmt.Println("Sample B", solve09("inputs/09_sample.txt", true))  // 24
	fmt.Println("B", solve09("inputs/09_input.txt", true))          // 1476550548
}

func solve09(filename string, checkInside bool) int {
	coords := parseCoords(filename)
	hSegs, vSegs := buildSegments(coords)

	type rect struct{ x1, y1, x2, y2, area int }
	var rects []rect
	for i, c1 := range coords {
		for _, c2 := range coords[i+1:] {
			x1, x2 := minmax(c1.x, c2.x)
			y1, y2 := minmax(c1.y, c2.y)
			rects = append(rects, rect{x1, y1, x2, y2, (x2 - x1 + 1) * (y2 - y1 + 1)})
		}
	}
	sort.Slice(rects, func(i, j int) bool { return rects[i].area > rects[j].area })

	for _, r := range rects {
		if !checkInside || isRectInside(r.x1, r.y1, r.x2, r.y2, hSegs, vSegs) {
			return r.area
		}
	}
	panic("no solution found")
}

func parseCoords(filename string) []coord {
	f, _ := os.Open(filename)
	defer f.Close()
	scanner := bufio.NewScanner(f)
	var coords []coord
	for scanner.Scan() {
		if line := scanner.Text(); line != "" {
			var c coord
			fmt.Sscanf(line, "%d,%d", &c.x, &c.y)
			coords = append(coords, c)
		}
	}
	return coords
}

func buildSegments(coords []coord) (hSegs, vSegs []seg) {
	for i, c1 := range coords {
		c2 := coords[(i+1)%len(coords)]
		if c1.y == c2.y {
			lo, hi := minmax(c1.x, c2.x)
			hSegs = append(hSegs, seg{c1.y, lo, hi})
		} else {
			lo, hi := minmax(c1.y, c2.y)
			vSegs = append(vSegs, seg{c1.x, lo, hi})
		}
	}
	return
}

func isRectInside(x1, y1, x2, y2 int, hSegs, vSegs []seg) bool {
	for _, h := range hSegs {
		if y1 < h.fixed && h.fixed < y2 && h.lo < x2 && x1 < h.hi {
			return false
		}
	}
	for _, v := range vSegs {
		if x1 < v.fixed && v.fixed < x2 && v.lo < y2 && y1 < v.hi {
			return false
		}
	}
	return true
}

func minmax(a, b int) (int, int) {
	if a > b {
		return b, a
	}
	return a, b
}

File: 2025_go/09_movie.py
#!/usr/bin/env python3

from itertools import combinations
from typing import NamedTuple


class Seg(NamedTuple):
    fixed: int
    lo: int
    hi: int


def build_segments(coords: list[tuple[int, int]]) -> tuple[list[Seg], list[Seg]]:
    h_segs, v_segs = [], []
    for (x1, y1), (x2, y2) in zip(coords, coords[1:] + coords[:1]):
        if y1 == y2:
            h_segs.append(Seg(y1, min(x1, x2), max(x1, x2)))
        else:
            v_segs.append(Seg(x1, min(y1, y2), max(y1, y2)))
    return h_segs, v_segs


def is_rect_inside(x1: int, y1: int, x2: int, y2: int, h_segs: list[Seg], v_segs: list[Seg]) -> bool:
    if any(y1 < h.fixed < y2 and h.lo < x2 and x1 < h.hi for h in h_segs):
        return False
    if any(x1 < v.fixed < x2 and v.lo < y2 and y1 < v.hi for v in v_segs):
        return False
    return True


def solve09(lines: list[str], b: bool) -> int:
    coords = [tuple(map(int, line.split(','))) for line in lines]
    h_segs, v_segs = build_segments(coords)

    rects = [
        (min(c1[0], c2[0]), min(c1[1], c2[1]), max(c1[0], c2[0]), max(c1[1], c2[1]))
        for c1, c2 in combinations(coords, 2)
    ]
    rects.sort(key=lambda r: (r[2] - r[0] + 1) * (r[3] - r[1] + 1), reverse=True)

    for x1, y1, x2, y2 in rects:
        if not b or is_rect_inside(x1, y1, x2, y2, h_segs, v_segs):
            return (x2 - x1 + 1) * (y2 - y1 + 1)

    raise ValueError("no solution found")


if __name__ == "__main__":
    with open("inputs/09_sample.txt") as f:
        s = [l for line in f if (l := line.strip())]
    with open("inputs/09_input.txt") as f:
        i = [l for line in f if (l := line.strip())]

    print(f"Sample A: {solve09(s, False)}")  # 50
    print(f"A:        {solve09(i, False)}")  # 4776100539
    print(f"Sample B: {solve09(s, True)}")  # 24
    print(f"B:        {solve09(i, True)}")  # 1476550548
